File: src/pydump.f90
! Temporary instrumentation for the Python-port feasibility spike (not part
! of the production solver). Dumps the static/geometric state EQdyna has
! built by the time the explicit time loop starts, so the Python port can
! load real Fortran-computed mesh/mass/shape-function data instead of
! re-deriving mesh4num/meshgen/assembleGlobalMass in Python for this spike.
subroutine pydump_state
    use globalvar
    implicit none
    integer (kind = 4) :: i, j, k, u

    u = 90001
    open(unit=u, file='pydump_header.txt', status='unknown')
    write(u,*) totalNumOfNodes
    write(u,*) totalNumOfElements
    write(u,*) totalNumOfEquations
    write(u,*) nen
    write(u,*) ned
    write(u,*) nftnd(1)
    write(u,*) ntotft
    write(u,*) nstep
    write(u,*) dt
    write(u,*) w
    write(u,*) rdampk
    write(u,*) rdampm
    write(u,*) kapa_hg
    write(u,*) R
    write(u,*) nPML
    write(u,*) vmaxPML
    write(u,*) (PMLb(i), i=1,8)
    write(u,*) grav
    write(u,*) C_elastic
    write(u,*) roumax
    write(u,*) rhow
    write(u,*) gamar
    write(u,*) slipRateThres
    write(u,*) xsource, ysource, zsource
    write(u,*) nucR, nucRuptVel, nucdtau0, nucT
    write(u,*) TPV
    write(u,*) C_nuclea
    write(u,*) nucfault
    write(u,*) friclaw
    write(u,*) timeElapsed
    close(u)

    open(unit=u, file='pydump_meshCoor.txt', status='unknown')
    do i = 1, totalNumOfNodes
        write(u,'(3(1x,e24.16e3))') meshCoor(1,i), meshCoor(2,i), meshCoor(3,i)
    enddo
    close(u)

    open(unit=u, file='pydump_conn.txt', status='unknown')
    do i = 1, totalNumOfElements
        write(u,'(8(1x,i8),1x,i3,5(1x,e24.16e3))') (nodeElemIdRelation(j,i), j=1,nen), elemTypeArr(i), &
            (mat(i,j), j=1,5)
    enddo
    close(u)

    open(unit=u, file='pydump_elemgeo.txt', status='unknown')
    do i = 1, totalNumOfElements
        write(u,'(1x,e24.16e3, 24(1x,e24.16e3), 6(1x,e24.16e3), 32(1x,e24.16e3))') eledet(i), &
            ((eleshp(j,k,i), j=1,3), k=1,8), (ss(j,i), j=1,6), ((phi(j,k,i), j=1,8), k=1,4)
    enddo
    close(u)

    open(unit=u, file='pydump_nodeinfo.txt', status='unknown')
    do i = 1, totalNumOfNodes
        write(u,'(1x,i2,1x,i10,12(1x,i10))') numOfDofPerNodeArr(i), eqNumStartIndexLoc(i), &
            (eqNumIndexArr(eqNumStartIndexLoc(i)+j), j=1,numOfDofPerNodeArr(i))
    enddo
    close(u)

    open(unit=u, file='pydump_nodalmass.txt', status='unknown')
    do i = 1, totalNumOfEquations
        write(u,'(1x,e24.16e3)') nodalMassArr(i)
    enddo
    close(u)

    open(unit=u, file='pydump_fnms.txt', status='unknown')
    do i = 1, totalNumOfNodes
        write(u,'(1x,e24.16e3)') fnms(i)
    enddo
    close(u)

    open(unit=u, file='pydump_fault.txt', status='unknown')
    do i = 1, nftnd(1)
        write(u,'(2(1x,i10),3(1x,e24.16e3),3(1x,e24.16e3),3(1x,e24.16e3),1x,e24.16e3,100(1x,e24.16e3))') &
            nsmp(1,i,1), nsmp(2,i,1), un(1,i,1),un(2,i,1),un(3,i,1), us(1,i,1),us(2,i,1),us(3,i,1), &
            ud(1,i,1),ud(2,i,1),ud(3,i,1), arn(i,1), (fric(j,i,1), j=1,100)
    enddo
    close(u)

    open(unit=u, file='pydump_v1.txt', status='unknown')
    do i = 1, totalNumOfEquations
        write(u,'(1x,e24.16e3)') v1(i)
    enddo
    close(u)

    ! Milestone 5 (standalone station matching): anonfs(1:3,1:numOfOnFaultStCount)
    ! maps on-fault station index i -> (fault-node-seq, xonfs column j, iFault);
    ! OffFaultStNodeIdIndex(1:2,1:numOfOffFaultStCount) maps off-fault station
    ! index -> (x4nds column i, matched nodeCount).
    open(unit=u, file='pydump_stations.txt', status='unknown')
    write(u,*) numOfOnFaultStCount
    write(u,*) numOfOffFaultStCount
    do i = 1, numOfOnFaultStCount
        write(u,'(3(1x,i10))') anonfs(1,i), anonfs(2,i), anonfs(3,i)
    enddo
    do i = 1, numOfOffFaultStCount
        write(u,'(2(1x,i10))') OffFaultStNodeIdIndex(1,i), OffFaultStNodeIdIndex(2,i)
    enddo
    close(u)
end subroutine pydump_state

! Per-step checkpoint (nt<=5) for first-divergence diagnosis: end-of-step
! nodal acceleration (nodalForceArr, already divided by mass), full velArr/
! dispArr, and the fault-node friction-state columns the port needs to
! reproduce (71-73 slip, 74-77 sliprate/cumslip, 47 final sliprate, 78-80
! traction, 20 RSF state (unused, friclaw=1), 23 normal-stress state).
subroutine pydump_step(step)
    use globalvar
    implicit none
    integer (kind = 4) :: step, i, j
    character(len=8) :: stmp
    write(stmp,'(i0)') step

    open(unit=90101, file='pydump_step'//trim(stmp)//'_accel.txt', status='unknown')
    do i = 1, totalNumOfEquations
        write(90101,'(1x,e24.16e3)') nodalForceArr(i)
    enddo
    close(90101)

    open(unit=90102, file='pydump_step'//trim(stmp)//'_veldisp.txt', status='unknown')
    do i = 1, totalNumOfNodes
        write(90102,'(6(1x,e24.16e3))') velArr(1,i),velArr(2,i),velArr(3,i), &
                                         dispArr(1,i),dispArr(2,i),dispArr(3,i)
    enddo
    close(90102)

    open(unit=90103, file='pydump_step'//trim(stmp)//'_fault.txt', status='unknown')
    do i = 1, nftnd(1)
        write(90103,'(11(1x,e24.16e3))') fric(71,i,1),fric(72,i,1),fric(73,i,1), &
            fric(74,i,1),fric(75,i,1),fric(76,i,1),fric(77,i,1),fric(47,i,1), &
            fric(78,i,1),fric(79,i,1),fric(80,i,1)
    enddo
    close(90103)
end subroutine pydump_step

File: python/eqdyna/pydump.py
"""pydump.py <- src/pydump.f90 (the reader for what it writes).

src/pydump.f90 is Fortran-side INSTRUMENTATION: called once before the time
loop, it dumps mesh, mass, shape-function, PML and fault-static state as
pydump_*.txt. This module reads those files back.

It is used by ONE thing -- python/run_parity.py, the per-step parity
diagnostic that answers "at which step does the port start to diverge from
Fortran's own internal state?". The solver itself never touches it:
eqdyna3d.build_solver_state builds every one of these fields from the case
inputs instead (readInputFiles.py + meshgen.py + assembleGlobalMass.py), so a
normal run needs no Fortran in the loop at all.

Kept separate from readInputFiles.py for that reason: that module reads the
CASE INPUTS the Fortran also reads, this one reads the Fortran's debug dump.
Merging them would put a diagnostic-only dependency in the solver's path.
"""
import numpy as np
import os


def _read_scalars(path):
    vals = []
    with open(path) as f:
        for line in f:
            vals.extend(line.split())
    return vals


def load(case_dir):
    h = _read_scalars(os.path.join(case_dir, 'pydump_header.txt'))
    it = iter(h)
    N = int(next(it)); E = int(next(it)); NEQ = int(next(it))
    nen = int(next(it)); ned = int(next(it))
    nftnd = int(next(it)); ntotft = int(next(it)); nstep = int(next(it))
    dt = float(next(it)); w = float(next(it)); rdampk = float(next(it))
    rdampm = float(next(it)); kapa_hg = float(next(it)); R = float(next(it))
    nPML = int(next(it)); vmaxPML = float(next(it))
    PMLb = np.array([float(next(it)) for _ in range(8)])
    grav = float(next(it)); C_elastic = int(next(it))
    roumax = float(next(it)); rhow = float(next(it)); gamar = float(next(it))
    slipRateThres = float(next(it))
    xsource, ysource, zsource = float(next(it)), float(next(it)), float(next(it))
    nucR, nucRuptVel, nucdtau0, nucT = [float(next(it)) for _ in range(4)]
    TPV = int(next(it)); C_nuclea = int(next(it)); nucfault = int(next(it))
    friclaw = int(next(it)); timeElapsed0 = float(next(it))

    meshCoor = np.loadtxt(os.path.join(case_dir, 'pydump_meshCoor.txt'))  # (N,3)
    conn_raw = np.loadtxt(os.path.join(case_dir, 'pydump_conn.txt'))      # (E,14)
    conn = conn_raw[:, 0:8].astype(np.int64) - 1  # 0-indexed node ids
    elemType = conn_raw[:, 8].astype(np.int64)
    mat = conn_raw[:, 9:14]  # vp,vs,rho,lam,miu

    eg = np.loadtxt(os.path.join(case_dir, 'pydump_elemgeo.txt'))
    eledet = eg[:, 0]
    eleshp = eg[:, 1:25].reshape(E, 8, 3)          # (E,8,3): stored ((eleshp(j,k)),j=1,3),k=1,8) -> j fastest
    ss = eg[:, 25:31]                               # (E,6)
    phi = eg[:, 31:63].reshape(E, 4, 8)             # stored ((phi(j,k)),j=1,8),k=1,4) -> j fastest -> phi[e,k-1,j-1]

    nodalMassArr = np.loadtxt(os.path.join(case_dir, 'pydump_nodalmass.txt'))  # (NEQ,)
    fnms = np.loadtxt(os.path.join(case_dir, 'pydump_fnms.txt'))              # (N,)
    v1 = np.loadtxt(os.path.join(case_dir, 'pydump_v1.txt'))                  # (NEQ,)

    ndof = np.zeros(N, dtype=np.int64)
    eq_ids = np.zeros((N, 12), dtype=np.int64)  # 0 = "no equation" sink
    with open(os.path.join(case_dir, 'pydump_nodeinfo.txt')) as f:
        for i, line in enumerate(f):
            parts = line.split()
            nd = int(parts[0])
            ndof[i] = nd
            eqs = np.array(parts[2:2 + nd], dtype=np.int64)
            eqs = np.where(eqs > 0, eqs, 0)  # -1 sentinel (fixed bnd) -> sink 0
            eq_ids[i, :nd] = eqs

    fq = np.loadtxt(os.path.join(case_dir, 'pydump_fault.txt'))  # (nftnd, 2+3+3+3+1+100)
    nsmp1 = fq[:, 0].astype(np.int64) - 1
    nsmp2 = fq[:, 1].astype(np.int64) - 1
    un = fq[:, 2:5]; us = fq[:, 5:8]; ud = fq[:, 8:11]
    arn = fq[:, 11]
    fric = fq[:, 12:112]  # columns 1..100 -> index 0..99

    return dict(N=N, E=E, NEQ=NEQ, nen=nen, ned=ned, nftnd=nftnd, ntotft=ntotft,
                nstep=nstep, dt=dt, w=w, rdampk=rdampk, rdampm=rdampm,
                kapa_hg=kapa_hg, R=R, nPML=nPML, vmaxPML=vmaxPML, PMLb=PMLb,
                grav=grav, C_elastic=C_elastic, roumax=roumax, rhow=rhow,
                gamar=gamar, slipRateThres=slipRateThres, xsource=xsource,
                ysource=ysource, zsource=zsource, nucR=nucR, nucRuptVel=nucRuptVel,
                nucdtau0=nucdtau0, nucT=nucT, TPV=TPV, C_nuclea=C_nuclea,
                nucfault=nucfault, friclaw=friclaw, meshCoor=meshCoor, conn=conn,
                elemType=elemType, mat=mat, eledet=eledet, eleshp=eleshp, ss=ss,
                phi=phi, nodalMassArr=nodalMassArr, fnms=fnms, v1_init=v1,
                ndof=ndof, eq_ids=eq_ids, nsmp1=nsmp1, nsmp2=nsmp2, un=un, us=us,
                ud=ud, arn=arn, fric_init=fric)
